### ISSUE TITLE
Missing Gradients When Using torch.autocast() and torch.no_grad() in Consecutive Model Forward Passes

### ISSUE THREAD

**Issue by Ignasijus** [pytorch/pytorch#142234]
### 🐛 Describe the bug

Performing multiple forward passes through the same model under a ``torch.autocast()`` context—where one pass is wrapped in ``torch.no_grad()`` and a subsequent pass is not—results in missing gradients. Specifically, parameters of the model do not receive gradients when autocast is enabled. However, if autocast is disabled, the gradients are computed as expected. 

To reproduce: 
Run the following code with autocast enabled and obsevre missing gradients.
Run the following code with autocast disabled and obsevre that gradients are present.

Additional info: if model has batchnorm layers, their gradients do not disappear. 

```python
import torch
import torch.nn as nn
import torch.optim as optim

class SimpleModel(nn.Module):
    def __init__(self, input_dim, output_dim):
        super(SimpleModel, self).__init__()
        self.fc = nn.Linear(input_dim, output_dim)

    def forward(self, x):
        x = self.fc(x)
        return x

device = "cuda" if torch.cuda.is_available() else "cpu"

model1 = SimpleModel(10, 20).to(device)
model2 = SimpleModel(20, 10).to(device)

optimizer = optim.Adam(list(model1.parameters()) + list(model2.parameters()), lr=1e-3)
scaler = torch.GradScaler('cuda')

# Dummy data
inputs = torch.randn(32, 10).to(device)
targets = torch.randint(0, 10, (32,)).to(device)

model1.train()
model2.train()

with torch.autocast(device_type=device, enabled=True):  # Enable or disable autocast
    embeddings = model1(inputs)
    
    with torch.no_grad():
        scores = model2(embeddings)
    
    outputs = model2(embeddings)
    loss = nn.CrossEntropyLoss()(outputs, targets)  

optimizer.zero_grad()
scaler.scale(loss).backward()

# Check if gradients are None for model2
for name, param in model2.named_parameters():
    print(f"{name} grad is None: {param.grad is None}")
```

### Versions

PyTorch version: 2.5.1+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Pro (10.0.19045 64-bit)
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.17 (main, Jul  5 2023, 20:47:11) [MSC v.1916 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19045-SP0
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: 555.99
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Name: 13th Gen Intel(R) Core(TM) i7-13700K
Manufacturer: GenuineIntel
Family: 198
Architecture: 9
ProcessorType: 3
DeviceID: CPU0
CurrentClockSpeed: 3400
MaxClockSpeed: 3400
L2CacheSize: 8192
L2CacheSpeed: None
Revision: None

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] onnx==1.13.1
[pip3] onnxruntime-openvino==1.13.1
[pip3] torch==2.5.1+cu124
[pip3] torch-geometric==2.6.1
[pip3] torchaudio==2.5.1+cu124
[pip3] torchvision==0.20.1+cu124
[conda] blas                      1.0                         mkl
[conda] mkl                       2021.4.0                 pypi_0    pypi
[conda] mkl-service               2.4.0            py39h2bbff1b_1
[conda] mkl_fft                   1.3.11           py39h827c3e9_0
[conda] mkl_random                1.2.8            py39hc64d2fc_0
[conda] numpy                     1.26.4           py39h055cbcc_0
[conda] numpy-base                1.26.4           py39h65a83cf_0
[conda] torch                     2.5.1+cu124              pypi_0    pypi
[conda] torch-geometric           2.6.1                    pypi_0    pypi
[conda] torchaudio                2.5.1+cu124              pypi_0    pypi
[conda] torchvision               0.20.1+cu124             pypi_0    pypi

cc @ezyang @albanD @gqchen @pearu @nikitaved @soulitzer @Varal7 @xmfan @mcarilli @ptrblck @leslie-fang-intel @jgong5
